src/importer.cpp

```diff
@@ -0,0 +1,484 @@
+                Variable::addEquivalence(importedComponentVariable, localModelVariable);
+            }
+        }
+        parent->replaceComponent(index, importedComponentCopy);
+
+        // Apply the rebased equivalence map onto the modified model.
+        applyEquivalenceMapToModel(rebasedMap, model);
+
+        // Copy over units used in imported component to this model.
+        std::map<std::string, std::string> unitsNamesToReplace;
+        for (const auto &u : requiredUnits) {
+            if (!model->hasUnits(u)) {
+                auto orignalName = u->name();
+                size_t count = 0;
+                while (!model->hasUnits(u) && model->hasUnits(u->name())) {
+                    auto name = u->name();
+                    name += "_" + convertToString(++count);
+                    u->setName(name);
+                }
+                model->addUnits(u);
+                if (orignalName != u->name()) {
+                    unitsNamesToReplace[orignalName] = u->name();
+                }
+            }
+        }
+        findAndReplaceComponentsCnUnitsNames(importedComponentCopy, unitsNamesToReplace);
+    }
+}
+
+void flattenComponentTree(const ComponentEntityPtr &parent, const ComponentPtr &component, size_t componentIndex)
+{
+    flattenComponent(parent, component, componentIndex);
+    auto flattenedComponent = parent->component(componentIndex);
+    for (size_t index = 0; index < flattenedComponent->componentCount(); ++index) {
+        auto c = flattenedComponent->component(index);
+        flattenComponentTree(flattenedComponent, c, index);
+    }
+}
+
+ModelPtr Importer::flatten(const ModelPtr &inModel)
+{
+    if (inModel->hasUnresolvedImports()) {
+        // Mimicking behaviour of previous flatten() function... not sure this is the way to go?
+        return inModel;
+    }
+    auto model = inModel->clone();
+
+    while (model->hasImports()) {
+        // Go through Units and instantiate any imported Units.
+        for (size_t index = 0; index < model->unitsCount(); ++index) {
+            auto u = model->units(index);
+            if (u->isImport()) {
+                auto importedUnits = u->importSource()->model()->units(u->importReference());
+                auto importedUnitsCopy = importedUnits->clone();
+                importedUnitsCopy->setName(u->name());
+                model->replaceUnits(index, importedUnitsCopy);
+            }
+        }
+
+        // Go through Components and instatiate any imported Components
+        for (size_t index = 0; index < model->componentCount(); ++index) {
+            auto c = model->component(index);
+            flattenComponentTree(model, c, index);
+        }
+    }
+
+    model->linkUnits();
+    return model;
+}
+
+size_t Importer::libraryCount()
+{
+    return mPimpl->mLibrary.size();
+}
+
+ModelPtr Importer::library(const std::string &key)
+{
+    if (mPimpl->mLibrary.count(key) > 0) {
```

Comment:
Replace `> 0` with `!= 0`. Same in line 453.

Reply:
There's no speed difference that I know of and the intent is clear? What am I missing?

Reply:
Just that this is what we do in in `utilities.cpp`. Also, I would expect the value to be either `0` or `1`, so I would expect `== 0` or `== 1`, but the rest of the code uses `!= 0`, hence my suggestion.

Reply:
5584669c